src/util/np_utils.py

```diff
@@ -4,3 +4,14 @@
 def normalized(v):
     norm = np.linalg.norm(v)
     return v / norm
+
+def perpendicular_2d(v):
```

Comment:
just flatten, permute, and reshape to the original dims instead of having 4 implementations of the same thing.